experimental/overhead_matching/swag/model/patch_embedding.py

```diff
@@ -0,0 +1,65 @@
+
+import common.torch.load_torch_deps
+import torch
+import torch.nn.functional as F
+import torchvision
+from dataclasses import dataclass
+from typing import Tuple
+
+
+@dataclass
+class WagPatchEmbeddingConfig:
+    patch_dims: Tuple[int, int]
+    num_aggregation_heads: int
+
+
+def feature_extraction(backbone, x):
+    features = backbone.features(x)
+    # If the dimension size is odd, then ceil_mode=True will pad by one on the right/bottom
+    max_pooled = F.max_pool2d(features, kernel_size=2, stride=2, padding=0, ceil_mode=True)
+    return max_pooled
+
+
+def compute_safa_input_dims(backbone: torch.nn.Module, patch_dims: Tuple[int, int]):
+    with torch.no_grad():
+        test_input = torch.empty(1, 3, *patch_dims)
+        result = feature_extraction(backbone, test_input)
+        return result.shape[-2] * result.shape[-1]
+
+
+class WagPatchEmbedding(torch.nn.Module):
+    def __init__(self, config: WagPatchEmbeddingConfig):
+        super().__init__()
+        self.backbone = torchvision.models.vgg19()
```

Comment:
why vgg19 instead of 16? (Page 64 https://www.proquest.com/docview/3132849999?pq-origsite=gscholar&fromopenview=true&sourcetype=Dissertations%20&%20Theses)

Reply:
from: https://arxiv.org/pdf/1409.1556